ENH: Function to calculate reference grids aligned with the coefficients

oesteban opened the pull request:
This is split off from #346, to focus only on this particular task, which is probably the most complicated aspect of the refactor (or the one we definitely want to get right).

oesteban committed:
enh: function that calculates the new reference grid

oesteban committed:
wip: add a test

effigies commented on sdcflows/utils/tools.py:
```diff
@@ -23,6 +23,71 @@
+        padding = tuple([padding] * 3)
+
+    underflow = min(extent_ijk.min(0) - padding, 0)
+    overflow = max(extent_ijk.max(0) + padding + 1 - ref_shape, 0)
```
This doesn't make sense to me. I think you probably want to be using `np.min`/`np.max` or `np.minimum`/`np.maximum`.

effigies commented on sdcflows/utils/tests/test_tools.py:
```diff
@@ -23,11 +23,50 @@
+    # Generate an example reference image
+    ref_data = np.zeros((20, 30, 40), dtype=np.float32)
+    ref_affine = np.eye(4)
+    ref_affine[:3, :3] = np.diag([2, 3, 4])  # Set zooms to (2, 3, 4)
```
```suggestion
    ref_affine = np.diag([2, 3, 4, 1])  # Set zooms to (2, 3, 4)
```

effigies commented on sdcflows/utils/tests/test_tools.py:
```diff
@@ -23,11 +23,50 @@
+    ob_data = np.zeros((25, 35, 45), dtype=np.float32)
+    ob_affine = np.eye(4)
+
+    # Rotate 90 degrees around x-axis
```
This test seems fine to test, but I would also add a 30deg test.

oesteban replied:
Yes, I was too lazy to pre-calculate what the shapes must result for several padding/factor combinations so opted for looking at the volume after interpolation. I have to check why there seems to be a substantial error when doing that. Once I figure that out, a 30 deg test (or whatever) could be added with parametrization (and tested with factors/paddings)

oesteban replied:
Added some parametric tests.

effigies commented on sdcflows/utils/tests/test_tools.py:
```diff
@@ -23,11 +23,50 @@
+
+    # Check output shape and zooms
+    assert out_img.shape == (26, 36, 46)
+    assert np.allclose(out_img.header.get_zooms()[:3], (0.5, 1.0, 1.0))
```
Some explanation would help. Why are the zooms 0.5x1x1? If we're upsampling by a factor of 4, shouldn't it be 0.5x0.75x1.0? And if we're padding by 1 in each direction, shouldn't the shape be (27, 37, 47)?

And what impact does the 90deg rotation have? Should it be 0.5x1.0x0.75 and (27, 47, 37)?

oesteban replied:
Sorry, I was still on it. I can give you a heads up when the test is ready.

oesteban committed:
fix: test minor items

Co-authored-by: Chris Markiewicz <effigies@gmail.com>

oesteban committed:
fix: address offset with factors

oesteban committed:
fix: finalize the implementation of a test

after long time scratching my head, it seems mango does not always
interpret orientation correctly. freeview seems to do so though.

effigies commented on sdcflows/utils/tools.py:
```diff
@@ -23,6 +23,70 @@
+        oblique.affine,
+        np.array(list(product((0, 1), repeat=3))) * (np.array(oblique.shape[:3]) - 1),
+    )
+    extent_ijk = apply_affine(np.linalg.inv(affine), extent)
```
I think it's clearer to give the complicated expression a name. We can keep it to two statements by composing the affines:

```suggestion
    corners = np.array(list(product((0, 1), repeat=3))) * (np.array(oblique.shape[:3]) - 1)
    extent_ijk = apply_affine(np.linalg.inv(affine) @ oblique.affine, corners)
```

effigies commented on sdcflows/utils/tools.py:
```diff
@@ -23,6 +23,70 @@
+    ).astype(int)
+    if np.any(underflow < 0) or np.any(overflow > 0):
+        # Add under/overflow voxels
+        ref_shape += np.abs(underflow) + overflow
```
Underflow is always non-positive.

```suggestion
        ref_shape += overflow - underflow
```

effigies commented on sdcflows/utils/tools.py:
```diff
@@ -23,6 +23,70 @@
+    hdr.set_qform(affine, qcode)
+
+    return in_reference.__class__(
+        np.zeros(ref_shape.astype(int), dtype=hdr.get_data_dtype()),
```
You don't need a dtype for a sampling reference. In fact, you don't need to allocate the whole array, just something that has a shape and can be made an array on demand.

```suggestion
        nb.fileslice.strided_scalar(ref_shape.astype(int)),
```

effigies reviewed (commented):
The function looks good. Minor comments. I will have a detailed look at the test when I can, but it won't be today.

mgxd commented on sdcflows/utils/tools.py:
```diff
@@ -23,6 +23,70 @@
 """Image processing tools."""
 
 
+def deoblique_and_zooms(in_reference, oblique, factor=4, padding=1):
```
may as well type it since you made that nice docstring
```suggestion
import nibabel as nb

def deoblique_and_zooms(
    in_reference: nb.spatialimages.SpatialImage, 
    oblique: nb.spatialimages.SpatialImage, 
    factor: int = 4, 
    padding: int = 1
):
```

effigies replied:
Note that `in_reference.header.get_qform()` is used, so `in_reference` needs to be `nb.Nifti1Image`, in case we ever want to type check this.

mgxd commented on sdcflows/utils/tools.py:
```diff
@@ -23,6 +23,70 @@
+        affine[:-1, -1] = apply_affine(affine, underflow)
+
+    # Make grid denser
+    if abs(1.0 - factor) > 1e-4:
```
Why this magic number? Should this be a configurable threshold?

oesteban replied:
Okay, I added a function argument for this. You're right the best practice would be to make it configurable.

oesteban committed:
Apply suggestions from code review

Co-authored-by: Mathias Goncalves <mathiasg@stanford.edu>
Co-authored-by: Chris Markiewicz <markiewicz@stanford.edu>
